Remove MDB_UNSIGNEDKEY, add a builder for Dbi, add MDB_INTEGERKEY comparators

at055612 opened the pull request:
Fixes #249 
Fixes #267 

Two fixes in one PR as both need each other.

benalexau commented:
Thanks @at055612 for this extensive PR.

Just looking over it, would you please remove formatting changes to files where they are not otherwise required, such as extra lines, spaces and hyphen-based section separators. Also the `src/main/resources` binaries are not needed as they are transitively available from the LmdbJava Native artifact.

I feel we should defer this PR until 0.10.0:

* There is unfortunately a risk of introducing regressions given the PR is very large at 6,500 lines and especially given it makes important changes to `Dbi`, `Comparator` and `Cursor` handling
* Generally a patch release should have minimal changes so that the stability of the minor version series is increasing with each successive patch
* While a patch release maintaining full API compatibility can reasonably mark a method as `@Deprecated` under semantic versioning principles, it is nevertheless much less surprising when incrementing the minor version number
* Introducing a builder pattern does represent evolution of how we recommend using the API. Again that would be something more ergonomically introduced with a new minor version number
* I have no objection to a 0.10.0 within a few days of 0.9.2 so that we can surface these changes (and consider #270 if it is ready by then). At least that way we know 0.9.2 is a definite reduction in regressions over 0.9.1 and offers a stable point-in-time regardless of any changes in 0.10.0

I have added the `gh-249-remove-unsignedkey` branch to the benchmarking suite and will update the published reports.

Does this seem reasonable to you?

at055612 commented:
@benalexau, releasing this in `0.10.0` is fine with me.  I agree with the reasons you have listed.

I have run the format plugin to sort out the whitespace and removed the separators for static inner classes (personally I think they improve readability but appreciate they are not everybody's cup of tea).  Not a fan of some of the decisions that the formatter is making, but formatting is incredibly subjective so having automated opinionated formatting does make life easier in a shared code base.

While this PR adds the Dbi builder, I decided to keep these methods for the really simple cases.  The builder can be used for everything else more complicated.  Hopefully a reasonable compromise.

```java
  public Dbi<T> openDbi(final byte[] name, final DbiFlagSet dbiFlagSet) {
  }

  public Dbi<T> openDbi(final String name, final DbiFlagSet dbiFlagSet) {
  }
``` 

I have removed the ByteUnits dependency as that repo is no longer being maintained and we were only using a fraction of it and only in tests.  I have replaced it with our own ByteUnit class and added new overloaded `setMapSize` methods to Env and Env.Builder to make it easier for users to specify larger sizes.

```java
    public Builder<T> setMapSize(final long mapSize, final ByteUnit byteUnit) {
    }
```

benalexau commented:
> I have run the format plugin to sort out the whitespace and removed the separators for static inner classes (personally I think they improve readability but appreciate they are not everybody's cup of tea). Not a fan of some of the decisions that the formatter is making, but formatting is incredibly subjective so having automated opinionated formatting does make life easier in a shared code base.

We just use the Spotify `fmt-maven-plugin` which implements the [Google Java Style](https://github.com/google/google-java-format) (which is prescriptive and has no configuration options). I used to use PMD and Checkstyle on LmdbJava but I found it problematic because (a) every update would introduce new warnings which had to be suppressed for no apparent benefit and (b) they reflect options which are debatable at best and impeded people who wanted to provide PRs. I think the code formatter's prescriptive and automated approach strikes the right balance.

> While this PR adds the Dbi builder, I decided to keep these methods for the really simple cases. The builder can be used for everything else more complicated. Hopefully a reasonable compromise.
> 
> ```java
>   public Dbi<T> openDbi(final byte[] name, final DbiFlagSet dbiFlagSet) {
>   }
> 
>   public Dbi<T> openDbi(final String name, final DbiFlagSet dbiFlagSet) {
>   }
> ```

The builders are such a significant ergonomic improvement and given we are targeting 1.0.0 it might be a good opportunity to remove those methods to encourage visibility of the new golden path.

> I have removed the ByteUnits dependency as that repo is no longer being maintained and we were only using a fraction of it and only in tests. I have replaced it with our own ByteUnit class and added new overloaded `setMapSize` methods to Env and Env.Builder to make it easier for users to specify larger sizes.

Excellent.

at055612 commented:
@benalexau
>The builders are such a significant ergonomic improvement and given we are targeting 1.0.0 it might be a good opportunity to remove those methods to encourage visibility of the new golden path.

Originally I had deprecated all of the `open(` methods so the only route for `Dbi` creation was the builder, however I then started to think that for the simplest cases where you just want a named db with no special options, it would be worthwhile to keep the simplest of the open methods. i.e. 

```java
        env.createDbi()
            .setDbName("foo")
            .withDefaultComparator()
            .setDbiFlags(DbiFlags.MDB_CREATE)
            .open();
``` 
vs
```java
env.open("foo", DbiFlags.MDB_CREATE)  // where MDB_CREATE is treated as a DbiFlagSet
```
The `open(byte[], DbiFlagSet)` variant can probably just be done via the builder as that seems like a less common use case.
Having this one `open(String, DbiFlagSet)` method may also reduce the amount of deprecation warnings people will see vs if it was deprecated.

I'm not too bothered either way so happy to go with whatever you decide.

at055612 commented:
@benalexau Works for me.  Builder it is. :+1:
